### PR TITLE
Add EC2 Platform Support for Java

### PR DESCRIPTION
**Pull request by harrryr**
*Issue #, if available:*
We need to create a new E2E test that tests changes for EC2 Instances managed by auto scaling groups as they have slightly different attributes. 

June 03, 2024 Update:
Java Artifacts have been released.

- Updated file name from appsignals -> application-signals
- Move Terraform and Validator code to Java folder to match Python
- Update testing-id

Test run: https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/9357092735

Test run after updating config: https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/9392804501

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/application-signals-java-e2e-ec2-asg-test.yml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+          --region ${{ inputs.aws-region }}
+          --account-id ${{ env.ACCOUNT_ID }}
+          --metric-namespace ${{ env.METRIC_NAMESPACE }}
+          --log-group ${{ env.LOG_GROUP_NAME }}
+          --service-name sample-application-${{ env.TESTING_ID }}
+          --remote-service-name sample-remote-application-${{ env.TESTING_ID }}
+          --instance-ami ${{ env.EC2_INSTANCE_AMI }}
+          --platform-info ec2-single-asg-${{ env.TESTING_ID }}
+          --instance-id ${{ env.INSTANCE_ID }}
+          --private-dns-name ${{ env.PRIVATE_DNS_NAME }}
```
Comment (majanjua-amzn) [inline .github/workflows/application-signals-java-e2e-ec2-asg-test.yml:233]: Is this being used? Can't see it anywhere, please remove if not.

File: .github/workflows/application-signals-java-e2e-ec2-canary-test.yml [hunk 2]
```diff
@@ -16,16 +16,30 @@ permissions:
...
+      caller-workflow-name: 'appsignals-e2e-ec2-canary-test'
+
+  e2e-ec2-asg-test:
+    strategy:
+      fail-fast: false
+      matrix:
+        aws-region: ['af-south-1','ap-east-1','ap-northeast-1','ap-northeast-2','ap-northeast-3','ap-south-1','ap-south-2','ap-southeast-1',
+                     'ap-southeast-2','ap-southeast-3','ap-southeast-4','ca-central-1','eu-central-1','eu-central-2','eu-north-1',
+                     'eu-south-1','eu-south-2','eu-west-1','eu-west-2','eu-west-3','il-central-1','me-central-1','me-south-1', 'sa-east-1',
+                     'us-east-1','us-east-2', 'us-west-1', 'us-west-2']
```
Comment (majanjua-amzn) [inline .github/workflows/application-signals-java-e2e-ec2-canary-test.yml:40]: Can we double check if any new regions have been added with the latest release?

File: terraform/ec2/default/amazon-cloudwatch-agent.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+  "agent": {
+    "debug": true,
+    "region": "$REGION"
+  },
+  "traces": {
+    "traces_collected": {
+      "app_signals": {
```
Comment (majanjua-amzn) [inline terraform/ec2/default/amazon-cloudwatch-agent.json:8]: This config has been changed, update it everywhere where applicable (only within the scope of this PR)
Also check if anything else has been changed, check public docs
```suggestion
      "application_signals": {
```

File: terraform/ec2/default/main.tf
Comment (majanjua-amzn) [inline terraform/ec2/default/main.tf:1]: Ty for making this improvement

File: validator/src/main/resources/expected-data-template/java/ec2/default/aws-sdk-call-log.mustache
Comment (majanjua-amzn) [inline validator/src/main/resources/expected-data-template/java/ec2/default/aws-sdk-call-log.mustache:1]: Ty for moving to `java` folder

File: terraform/ec2/default/amazon-cloudwatch-agent.json (hunk 3 above)
Comment (harrryr) [reply terraform/ec2/default/amazon-cloudwatch-agent.json:8]: Done

Comment (harrryr) [reply terraform/ec2/default/amazon-cloudwatch-agent.json:8]: Also removed `enabled`

File: .github/workflows/application-signals-java-e2e-ec2-canary-test.yml (hunk 2 above)
Comment (harrryr) [reply .github/workflows/application-signals-java-e2e-ec2-canary-test.yml:40]: Yeah no new regions

File: .github/workflows/application-signals-java-e2e-ec2-asg-test.yml (hunk 1 above)
Comment (harrryr) [reply .github/workflows/application-signals-java-e2e-ec2-asg-test.yml:233]: You mean `PRIVATE_DNS_NAME`? Yeah it's being used

Comment (harrryr) [reply .github/workflows/application-signals-java-e2e-ec2-asg-test.yml:233]: Ex: https://github.com/aws-observability/aws-application-signals-test-framework/pull/72/files#diff-ed392d4a48e7a1e3c84a3105f421b0b15a8633134f641de91f910cac3c6099c2R23

File: .github/workflows/application-signals-java-e2e-ec2-asg-test.yml [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            fi
+
+            # If the deployment_failed is still 0, then the terraform deployment succeeded and now try to connect to the endpoint.
+            # Attempts to connect will be made for up to 10 minutes
+            if [ $deployment_failed -eq 0 ]; then
+              echo "Attempting to connect to the endpoint"
+              main_service_instance_id=$(aws autoscaling describe-auto-scaling-groups --auto-scaling-group-names ec2-single-asg-${{ env.TESTING_ID }} --region ${{ inputs.aws-region }} --query "AutoScalingGroups[].Instances[0].InstanceId" --output text)
+              main_service_public_ip=$(aws ec2 describe-instances --instance-ids $main_service_instance_id --region ${{ inputs.aws-region }} --query "Reservations[].Instances[].PublicIpAddress" --output text)
+              main_service_sample_app_endpoint=http://$main_service_public_ip:8080
+              echo "The main service endpoint is $main_service_sample_app_endpoint"
```
[the remark this answers is not in the document]
Comment (harrryr) [reply .github/workflows/application-signals-java-e2e-ec2-asg-test.yml]: Moved them up to Terraform Deployment Step

File: terraform/ec2/default/main.tf [hunk 5]
```diff
@@ -104,36 +104,41 @@ resource "null_resource" "main_service_setup" {
...
-      "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://localhost:4315 \\",
-      "OTEL_RESOURCE_ATTRIBUTES=service.name=sample-application-${var.test_id} \\",
-      "nohup java -jar main-service.jar &> nohup.out &",
+      aws s3 cp ${var.sample_app_jar} ./main-service.jar
+
+      JAVA_TOOL_OPTIONS=' -javaagent:/home/ec2-user/adot.jar' \
+      OTEL_METRICS_EXPORTER=none \
+      OTEL_LOGS_EXPORT=none \
+      OTEL_AWS_APP_SIGNALS_ENABLED=true \
+      OTEL_AWS_APP_SIGNALS_EXPORTER_ENDPOINT=http://localhost:4316/v1/metrics \
```
Comment (majanjua-amzn) [inline terraform/ec2/default/main.tf:132]: Supposed to be like this, double check docs to see what the recommendation for may release is.
```suggestion
      OTEL_AWS_APPLICATION_SIGNALS_ENABLED=true \
      OTEL_AWS_APPLICATION_SIGNALS_EXPORTER_ENDPOINT=http://localhost:4316/v1/metrics \
```

File: terraform/ec2/asg/main.tf [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    ${var.get_adot_jar_command}
+
+    # Get and run the sample application with configuration
+    aws s3 cp ${var.sample_app_jar} ./main-service.jar
+
+    JAVA_TOOL_OPTIONS=' -javaagent:/adot.jar' \
+    OTEL_METRICS_EXPORTER=none \
+    OTEL_LOGS_EXPORT=none \
+    OTEL_AWS_APP_SIGNALS_ENABLED=true \
+    OTEL_AWS_APP_SIGNALS_EXPORTER_ENDPOINT=http://localhost:4316/v1/metrics \
```
Comment (majanjua-amzn) [inline terraform/ec2/asg/main.tf:121]: Same here

File: terraform/ec2/asg/amazon-cloudwatch-agent.json [hunk 7]
```diff
@@ -5,16 +5,12 @@
   },
   "traces": {
     "traces_collected": {
-      "app_signals": {
-        "enabled": true
-      }
+      "application_signals": {}
```
Comment (majanjua-amzn) [inline terraform/ec2/asg/amazon-cloudwatch-agent.json:8]: Just confirming, we don't see `enabled` here anymore?

Comment (harrryr) [reply terraform/ec2/asg/amazon-cloudwatch-agent.json:8]: Yes

File: terraform/ec2/default/main.tf (hunk 5 above)
Comment (harrryr) [reply terraform/ec2/default/main.tf]: Updated based on here: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Application-Signals-Enable-EC2.html

File: terraform/ec2/asg/main.tf (hunk 6 above)
Comment (harrryr) [reply terraform/ec2/asg/main.tf]: Done

**Pull request closed without merging**